content/blog/2019-12-18-novm/index.md

```diff
@@ -0,0 +1,317 @@
+In this world, we can still ask our OS for memory via `malloc`; however it returns
+back to us a physically contiguous memory region (rather than virtually contiguous).
+For the large memory applications described above that manage their own memory,
+they would likely start by `malloc`-ing most of the computer's physical memory
+and then never calling `malloc` again. Little would change for such programs
+(except that the spatial locality assumptions their designers had originally
+made about memory layout are more likely to reflect reality).
+
+However, programs which request new allocations throughout their lifetimes
+may no longer be able to execute correctly. Since `malloc` returns a physical memory region,
+the OS needs to find a large enough space inside the memory to allocate. Due to the
+presence of [fragmentation](https://en.wikipedia.org/wiki/Fragmentation_(computing)),
+it is possible that no such region exists. In that case, `malloc` returns `0` and,
+in all likelihood, the program explodes.
+
+Remember that such fragmentation was present
+with virtual addressing as well, but the OS could stitch together various fragmented segments
+to form a single virtual allocation. Therefore, programs should strive to allocate memory in fixed-size chunks;
+essentially, they should assume that the OS can only allocate them pages of physical memory
+and it's _their job to stripe datastructures across them_.
+
+# Experimental Setup
+
+To evaluate the impact of software changes required in lieu of virtual addressing,
+we ran experiments with the following configurations. First, we ran all of our tests
+on a computer with 8 Intel i7-7700 CPUs clocked at 3.60GHz, with 32GB of physical memory, running Ubuntu 16.04.
+Secondly, we followed the [guidlines](https://llvm.org/docs/Benchmarking.html) provided by llvm to reduce
+variance; in particular, every test was executed on a single, isolated processor core.
+While we ran all of our tests ten times and report averages of our measurements, with
+this setup we observed very little variance with typically less than 0.01% standard deviation.
+Finally, we assumed that some reasonable amount of stack could be pre-allocated contiguously,
+even on a physically addressed machine. We chose 32KB since that was approximately the smallest
+sized stack required to execute the benchmarks normally.
+
+Unfortunately, we could not actually execute any tests using physical addressing,
+since there is no reliable method for allocating physical memory in user space.
+While there are several [proposals](https://blog.linuxplumbersconf.org/2017/ocw/system/presentations/4669/original/Support%20user%20space%20POSIX%20conformant%20contiguous__v1.00.pdf) for how to implement these
+features, they aren't currently supported in Linux. Therefore, our results
+are overhead measurements that represent worst-case performance; we don't actually
+expect any of our tests to result in speedups.
```

Comment:
It does seem like you could get part of the way there by allocating a few hugepages and then "mallocing" from that pool…